src/logger.hpp

```diff
@@ -0,0 +1,164 @@
+using Entry = std::pair<LoggerLevel, std::string>;
+using Delegated = std::function<void(int, std::string)>;
+
+/*
+ * The idea here is to one day implement a functionality to delegate the logs
+ * to the mobile OS, for that we got two specialized behaviour that we can
+ * inject at compile time or at runtime.
+ *
+ */
+
+#if REALM_ANDROID
+class AndroidLogger {
+    std::map<LoggerLevel, android_LogPriority> map_android_log_level{
+        {LoggerLevel::all, ANDROID_LOG_VERBOSE},    {LoggerLevel::info, ANDROID_LOG_INFO},
+        {LoggerLevel::trace, ANDROID_LOG_DEFAULT},  {LoggerLevel::debug, ANDROID_LOG_DEBUG},
+        {LoggerLevel::detail, ANDROID_LOG_VERBOSE}, {LoggerLevel::warn, ANDROID_LOG_WARN},
+        {LoggerLevel::error, ANDROID_LOG_ERROR},    {LoggerLevel::fatal, ANDROID_LOG_FATAL},
+        {LoggerLevel::off, ANDROID_LOG_SILENT},
+    };
+
+    void print(Entry& entry)
+    {
+        auto android_log_level = map_android_log_level[entry.first];
+        __android_log_print(android_log_level, "realm", "%s", entry.second.c_str());
+    }
+};
+#endif
+
+#ifdef __APPLE__
+class IOSLogger {
+    void print() {}
+};
+#endif
+
+class SyncLogger : public realm::util::RootLogger {
+public:
+    void delegate(Delegated&& delegate)
+    {
+        m_scheduler->set_notify_callback([this, delegate = std::move(delegate)] {
+            while (!m_log_queue.empty()) {
+                /*
+                 *  Extracting and delegating log entries.
+                 */
+                Entry entry;
+                {
+                    std::lock_guard<std::mutex> lock(m_mutex);
+                    if (m_log_queue.empty())
+                        return;
+                    entry = m_log_queue.front();
+                    m_log_queue.pop();
+                }
+
+                delegate(static_cast<int>(entry.first), entry.second);
+            }
+        });
+    }
+
+protected:
+    void do_log(LoggerLevel level, std::string message)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        // TODO we are coupling core with JS here, change to string use hashmap
+        // map_level.
+        auto entry = std::make_pair(level, message);
+
+        m_log_queue.push(entry);
+        m_scheduler->notify();
+    }
+
+private:
+    std::queue<Entry> m_log_queue;
+    Delegated loggerDelegate;
+    std::shared_ptr<realm::util::Scheduler> m_scheduler = realm::util::Scheduler::make_default();
+    std::mutex m_mutex;
+};
+
+class LoggerFactory : public realm::SyncLoggerFactory {
+public:
+    LoggerFactory(Delegated&& logs_fn)
+        : logs_fn{logs_fn}
+    {
+    }
+
+    std::unique_ptr<realm::util::Logger> make_logger(realm::util::Logger::Level level)
+    {
+        auto logger = std::make_unique<SyncLogger>();
+
+        logger->set_level_threshold(level);
+        logger->delegate(std::move(logs_fn));
+
+        return logger;
+    }
+
+private:
+    Delegated logs_fn;
+};
+
+class Logger {
```

Comment:
[Nitpick]  If you decide to rename `LoggerFactory`, probably this class should be renamed to reflect the new name as well.